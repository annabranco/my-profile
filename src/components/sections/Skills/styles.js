import styled, { css } from 'styled-components';
import { rgba } from 'polished';
import {
  colorGreenBright,
  colorWhite,
  colorBlack,
  colorBlueWater,
  fontTitleAlt,
  fontTitle,
  colorYellowBright
} from '../../../styles/theme';
import { isDesktop } from '../../../utils/device';

export const Logo = styled.img`
  margin: 0 auto;
  height: 30%;
  width: 30%;
`;
Logo.displayName = 'Logo';

export const Name = styled.h4`
  margin-bottom: 5px;
  font-family: ${fontTitle};
  font-size: 1rem;
  font-weight: bold;
  color: ${colorBlack};
`;
Name.displayName = 'Name';

export const SkillsArea = styled.div`
  display: grid;
  grid-template-columns: repeat(4, calc(100% / 4));
  grid-gap: 20px 5px;
  border-radius: 30px;
  background-image: linear-gradient(
    ${rgba(colorGreenBright, 0.1)},
    ${rgba(colorWhite, 0.2)}
  );
  height: 100%;
  width: 100vw;
  padding: 45px 40px 80px;
  overflow: hidden;

  @media all and (min-width: 768px) {
    grid-template-columns: repeat(4, calc(100% / 4 - 15px));
    grid-gap: 20px;
    height: 600px;
    width: 90vw;
    transition: all ease 2s;
  }
`;
SkillsArea.displayName = 'Skills Area';

export const SkillsGrid = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  @media all and (min-width: 768px) {
    display: grid;
    grid-template-columns: repeat(2, 60px);
    grid-gap: 20px;
  }
`;
SkillsGrid.displayName = 'Skills Grid';

export const SkillGroup = styled.div`
  max-width: 280px;
  border-radius: 10%;
  opacity: 0;

  &:hover {
    background: ${rgba(colorYellowBright, 0.1)};
  }
  transition: transform 2s ease, opacity 1s ease;

  ${({ position }) => {
    if (position === 1 || position === 5) {
      return css`
        transform: translate(-300px, 200px) rotate(64deg);
      `;
    }
    if (position === 2 || position === 6) {
      return css`
        transform: translate(-300px, 500px) rotate(10deg);
      `;
    }
    if (position === 3 || position === 7) {
      return css`
        transform: translate(300px, 500px) rotate(-10deg);
      `;
    }
    if (position === 4 || position === 8) {
      return css`
        transform: translate(300px, 200px) rotate(-40deg);
      `;
    }
    return null;
  }};

  ${({ isVisible }) =>
    isVisible &&
    css`
      transform: translate(0, 0);

      margin-top: 0;
      opacity: 1;
    `}

  @media all and (min-width: 768px) {
    margin: 20px 0 10px;
    padding-bottom: 10px;
  }
`;
SkillGroup.displayName = 'Skill Group';

export const SkillsInsideGroup = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  width: 100%;

  @media all and (min-width: 768px) {
    flex-direction: row;
  }
`;
SkillsInsideGroup.displayName = 'Skills Inside Group';

export const SkillItem = styled.div`
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin: 0 auto 10px;

  ${({ isLastElementAlone }) =>
    isDesktop &&
    isLastElementAlone &&
    css`
      &:last-of-type {
        grid-column: span 2;
      }
    `}
`;
SkillItem.displayName = 'Skill Item';

export const Title = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  border: 1px solid ${colorBlack};
  border-radius: 10px 0;
  margin: 0 auto 15px;
  background-image: linear-gradient(
    ${rgba(colorWhite, 0.2)},
    ${rgba(colorBlueWater, 0.5)}
  );
  min-height: ${({ group }) =>
    !isDesktop && group !== 'Frontend' ? '60px' : 'auto'};
  width: 100%;
  padding: 5px 0;
  text-decoration: none;
  font-family: ${fontTitleAlt};
  text-align: center;
`;
Title.displayName = 'Title';
