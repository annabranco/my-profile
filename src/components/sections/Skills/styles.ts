import * as React from 'react';
import styled, { css } from 'styled-components';
import { rgba } from 'polished';
import {
  colorBlack,
  colorBlueWater,
  colorGreenBright,
  colorWhite,
  colorYellowBright,
  fontTitle,
  fontTitleAlt
} from '../../../styles/theme';

export const Logo = styled.img`
  margin: 0 auto;
  height: 30%;
  width: 30%;
`;
Logo.displayName = 'Logo';

export const Name = styled.h4`
  margin-bottom: 5px;
  font-family: ${fontTitle};
  font-size: 1rem;
  font-weight: bold;
  color: ${colorBlack};
`;
Name.displayName = 'Name';

export const SkillsArea = styled.div`
  display: grid;
  grid-template-columns: repeat(4, calc(100% / 4));
  grid-gap: 20px 5px;
  justify-content: center;
  align-items: center;
  border-radius: 30px;
  background-image: linear-gradient(
    ${rgba(colorGreenBright, 0.1)},
    ${rgba(colorWhite, 0.2)}
  );
  height: 95%;
  width: 100vw;
  padding: 45px 40px 80px;
  overflow: hidden;

  @media all and (min-width: 768px) {
    grid-template-columns: repeat(4, calc(100% / 4 - 15px));
    grid-gap: 20px;
    height: 600px;
    width: 90vw;
    transition: all ease 2s;
  }
`;
SkillsArea.displayName = 'Skills Area';

export const SkillsGrid = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  @media all and (min-width: 768px) {
    display: grid;
    grid-template-columns: repeat(2, 60px);
    grid-gap: 20px;
  }
`;
SkillsGrid.displayName = 'Skills Grid';

interface SkillGroupProps extends React.ComponentPropsWithoutRef<'div'> {
  isVisible?: boolean;
  position?: number;
}

export const SkillGroup = styled.div<SkillGroupProps>`
  max-width: 280px;
  border-radius: 10%;
  opacity: 0;

  &:hover {
    background: ${rgba(colorYellowBright, 0.1)};
  }
  transition: transform 2s ease, opacity 1s ease;

  ${({ position }) => {
    if (position === 1 || position === 5) {
      return css`
        transform: translate(-300px, 200px) rotate(64deg);
      `;
    }
    if (position === 2 || position === 6) {
      return css`
        transform: translate(-300px, 500px) rotate(10deg);
      `;
    }
    if (position === 3 || position === 7) {
      return css`
        transform: translate(300px, 500px) rotate(-10deg);
      `;
    }
    if (position === 4 || position === 8) {
      return css`
        transform: translate(300px, 200px) rotate(-40deg);
      `;
    }
    return null;
  }};

  ${({ isVisible }) =>
    isVisible &&
    css`
      transform: translate(0, 0);

      margin-top: 0;
      opacity: 1;
    `}

  @media all and (min-width: 768px) {
    margin: 20px 0 10px;
    padding-bottom: 10px;
  }
`;
SkillGroup.displayName = 'Skill Group';

export const SkillsInsideGroup = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  width: 100%;

  @media all and (min-width: 768px) {
    flex-direction: row;
  }
`;
SkillsInsideGroup.displayName = 'Skills Inside Group';

interface SkillItemProps extends React.ComponentPropsWithoutRef<'div'> {
  group?: string;
  isLastElementAlone?: boolean;
}

export const SkillItem = styled.div<SkillItemProps>`
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin: ${({ group }) =>
    group === 'Frontend' ? '0 auto 10px' : '0 auto 2px'};

  ${({ isLastElementAlone }) =>
    isLastElementAlone &&
    css`
      @media all and (min-width: 768px) {
        &:last-of-type {
          grid-column: span 2;
        }
      }
    `}

  @media all and (min-width: 768px) {
    margin: 0 auto 10px;
  }
`;
SkillItem.displayName = 'Skill Item';

interface TitleProps extends React.ComponentPropsWithoutRef<'div'> {
  group?: string;
}

export const Title = styled.div<TitleProps>`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  border: 1px solid ${colorBlack};
  border-radius: 10px 0;
  margin: 0 auto 5px;
  background-image: linear-gradient(
    ${rgba(colorWhite, 0.2)},
    ${rgba(colorBlueWater, 0.5)}
  );
  min-height: ${({ group }) => (group !== 'Frontend' ? '60px' : 'auto')};
  width: 100%;
  padding: 5px 0;
  text-decoration: none;
  font-family: ${fontTitleAlt};
  text-align: center;

  @media all and (min-width: 768px) {
    min-height: auto;
    margin: 0 auto 15px;
  }
`;
Title.displayName = 'Title';
