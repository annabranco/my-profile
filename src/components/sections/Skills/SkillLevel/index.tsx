import { ReactElement } from 'react';
import { number, string } from 'prop-types';
import { Star, StarsWrapper } from './styles';

interface Props {
  level: number;
  skill: string;
}

const SkillLevel = ({ level, skill }: Props): ReactElement => {
  const stars: '*'[] = new Array(level).fill('*');

  return (
    <StarsWrapper data-e2e-id={`skill-${skill}_starsWrapper`}>
      {stars.map((_, index) => (
        <Star
          aria-hidden
          data-e2e-id={`skill-${skill}_star`}
          className="fas fa-star"
          key={`${skill}-${level}-${index}`} // eslint-disable-line react/no-array-index-key
        />
      ))}
    </StarsWrapper>
  );
};

SkillLevel.propTypes = {
  skill: string.isRequired,
  level: number.isRequired
};

export default SkillLevel;
