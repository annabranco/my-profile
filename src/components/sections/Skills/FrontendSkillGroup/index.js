import React from 'react';
import { bool } from 'prop-types';
import SkillLevel from '../SkillLevel';
import { FrontendGroup, Logo, LogoWrapper, SkillsWrapper } from './styles';
import { Name, SkillItem, Title } from '../styles';
import { skillGroupsPropType } from '../../../../types';
import { isDesktop } from '../../../../utils/device';

const FrontendSkillGroup = ({ details, visible }) => (
  <FrontendGroup visible={visible}>
    <Title group={details.name}>
      <h3>{details.name}</h3>
    </Title>
    <SkillsWrapper>
      {details.skills.map(skill => {
        if (!isDesktop && skill.extended) {
          return null;
        }
        return (
          <SkillItem key={`${skill.order}-${skill.skill}`}>
            <Name>{skill.skill}</Name>

            {skill.logo && (
              <LogoWrapper>
                <Logo src={skill.logo} alt={skill.skill} name={skill.skill} />
              </LogoWrapper>
            )}

            <SkillLevel level={skill.level} skill={skill.skill} />
          </SkillItem>
        );
      })}
    </SkillsWrapper>
  </FrontendGroup>
);

FrontendSkillGroup.propTypes = {
  details: skillGroupsPropType.isRequired,
  visible: bool.isRequired
};

export default FrontendSkillGroup;
