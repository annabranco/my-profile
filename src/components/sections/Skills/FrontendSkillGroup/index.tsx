import { ReactElement } from 'react';
import { bool } from 'prop-types';
import { ISkillGroups } from '../../../../types/interfaces';
import SkillLevel from '../SkillLevel';
import { isDesktop } from '../../../../utils/device';
import { skillGroupsPropType } from '../../../../types/propTypes';
import { FrontendGroup, Logo, LogoWrapper, SkillsWrapper } from './styles';
import { Name, SkillItem, Title } from '../styles';

interface Props {
  details: ISkillGroups;
  visible: boolean;
}

const FrontendSkillGroup = ({ details, visible }: Props): ReactElement => (
  <FrontendGroup visible={visible}>
    <Title group={details.name}>
      <h3>{details.name}</h3>
    </Title>
    <SkillsWrapper>
      {details.skills.map(skill => {
        if (!isDesktop && skill.extended) {
          return null;
        }
        return (
          <SkillItem group={details.name} key={`${skill.order}-${skill.skill}`}>
            <Name>{skill.skill}</Name>

            {skill.logo && (
              <LogoWrapper>
                <Logo src={skill.logo} alt={skill.skill} name={skill.skill} />
              </LogoWrapper>
            )}

            <SkillLevel level={skill.level} skill={skill.skill} />
          </SkillItem>
        );
      })}
    </SkillsWrapper>
  </FrontendGroup>
);

FrontendSkillGroup.propTypes = {
  details: skillGroupsPropType.isRequired,
  visible: bool.isRequired
};

export default FrontendSkillGroup;
