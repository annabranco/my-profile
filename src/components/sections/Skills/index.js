import React, { Fragment } from 'react';
import { useSelector } from 'react-redux';
import { instanceOf } from 'prop-types';
import FrontendSkillGroup from './FrontendSkillGroup';
import SkillLevel from './SkillLevel';
import { skillsSelector } from '../../../redux/selectors';
import {
  SKILLS_FIRST_ROW,
  SKILLS_SECOND_ROW,
  SKILLS_THIRD_ROW
} from '../../../constants';
import {
  SkillsArea,
  Title,
  SkillsGrid,
  SkillItem,
  Name,
  SkillGroup,
  SkillsInsideGroup,
  Logo
} from './styles';
import { isDesktop } from '../../../utils/device';

const FIRST_ROW_ITEMS = 'Frontend';
const SECOND_ROW_ITEMS = new Set([1, 2, 3, 4]);
const THIRD_ROW_ITEMS = new Set([5, 6, 7, 8]);

const Skills = ({ cuePointsActivated }) => {
  const skills = useSelector(skillsSelector);

  const isVisible = identificator => {
    if (
      cuePointsActivated.has(SKILLS_FIRST_ROW) &&
      FIRST_ROW_ITEMS.includes(identificator)
    ) {
      return true;
    }
    if (
      cuePointsActivated.has(SKILLS_SECOND_ROW) &&
      SECOND_ROW_ITEMS.has(identificator)
    ) {
      return true;
    }
    if (
      cuePointsActivated.has(SKILLS_THIRD_ROW) &&
      THIRD_ROW_ITEMS.has(identificator)
    ) {
      return true;
    }
    return false;
  };

  return (
    <SkillsArea>
      {skills.map((skillObject, index) => (
        <Fragment key={skillObject.name}>
          {skillObject.name === 'Frontend' ? (
            <FrontendSkillGroup
              details={skillObject}
              visible={isVisible(skillObject.name)}
            />
          ) : (
            <SkillGroup isVisible={isVisible(index)} position={index}>
              <Title>
                <h3>{skillObject.name}</h3>
              </Title>
              <SkillsInsideGroup>
                {isDesktop && <Logo src={skillObject.logo} alt="" />}
                <SkillsGrid>
                  {skillObject.skills.map(skill => {
                    if (skill.extended && !isDesktop) {
                      return null;
                    }
                    return (
                      <SkillItem
                        key={skill.skill}
                        isLastElementAlone={skillObject.skills.length % 2 !== 0}
                      >
                        <Name>{skill.skill}</Name>
                        <SkillLevel level={skill.level} skill={skill.skill} />
                      </SkillItem>
                    );
                  })}
                </SkillsGrid>
              </SkillsInsideGroup>
            </SkillGroup>
          )}
        </Fragment>
      ))}
    </SkillsArea>
  );
};

Skills.propTypes = {
  cuePointsActivated: instanceOf(Set).isRequired
};

export default Skills;
