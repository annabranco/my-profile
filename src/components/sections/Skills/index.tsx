import React, { Fragment, ReactElement } from 'react';
import { useSelector } from 'react-redux';
import { instanceOf } from 'prop-types';
import { AppState, SkillGroupsType } from '../../../types/interfaces';
import { skillsSelector } from '../../../redux/selectors';
import { isDesktop } from '../../../utils/device';
import FrontendSkillGroup from './FrontendSkillGroup';
import SkillLevel from './SkillLevel';
import {
  SKILLS_FIRST_ROW,
  SKILLS_SECOND_ROW,
  SKILLS_THIRD_ROW
} from '../../../constants';
import {
  Logo,
  Name,
  SkillGroup,
  SkillItem,
  SkillsArea,
  SkillsGrid,
  SkillsInsideGroup,
  Title
} from './styles';

const FIRST_ROW_ITEMS = 'Frontend';
const SECOND_ROW_ITEMS: Set<number> = new Set([1, 2, 3, 4]);
const THIRD_ROW_ITEMS: Set<number> = new Set([5, 6, 7, 8]);

interface Props {
  cuePointsActivated: Set<string>;
}

const Skills = ({ cuePointsActivated }: Props): ReactElement => {
  const skills: AppState['skills'] = useSelector(skillsSelector);

  const isVisible = (identificator: string | number): boolean => {
    if (typeof identificator === 'string') {
      if (
        cuePointsActivated.has(SKILLS_FIRST_ROW) &&
        FIRST_ROW_ITEMS.includes(identificator)
      ) {
        return true;
      }
    } else {
      if (
        cuePointsActivated.has(SKILLS_SECOND_ROW) &&
        SECOND_ROW_ITEMS.has(identificator)
      ) {
        return true;
      }
      if (
        cuePointsActivated.has(SKILLS_THIRD_ROW) &&
        THIRD_ROW_ITEMS.has(identificator)
      ) {
        return true;
      }
    }
    return false;
  };

  return (
    <SkillsArea>
      {skills.map((skillObject: SkillGroupsType, index: number) => (
        <Fragment key={skillObject.name}>
          {skillObject.name === 'Frontend' ? (
            <FrontendSkillGroup
              details={skillObject}
              visible={isVisible(skillObject.name)}
            />
          ) : (
            <SkillGroup isVisible={isVisible(index)} position={index}>
              <Title>
                <h3>{skillObject.name}</h3>
              </Title>
              <SkillsInsideGroup>
                {isDesktop && <Logo src={skillObject.logo} alt="" />}
                <SkillsGrid>
                  {skillObject.skills.map(skill => {
                    if (skill.extended && !isDesktop) {
                      return null;
                    }
                    return (
                      <SkillItem
                        key={skill.skill}
                        isLastElementAlone={skillObject.skills.length % 2 !== 0}
                      >
                        <Name>{skill.skill}</Name>
                        <SkillLevel level={skill.level} skill={skill.skill} />
                      </SkillItem>
                    );
                  })}
                </SkillsGrid>
              </SkillsInsideGroup>
            </SkillGroup>
          )}
        </Fragment>
      ))}
    </SkillsArea>
  );
};

Skills.propTypes = {
  cuePointsActivated: instanceOf(Set).isRequired
};

export default Skills;
