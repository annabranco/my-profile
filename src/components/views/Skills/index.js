import React from 'react';
import { arrayOf } from 'prop-types';
import FrontendSkillGroup from '../FrontendSkillGroup';
import SkillLevel from '../SkillLevel';
import { skillGroupsPropType } from '../../../types';
import {
  Sidebar,
  SkillGroupTitle,
  SkillsGrid,
  SkillWrapper,
  SkillName,
  SkillGroup,
  SkillsList,
  SkillsLogo
} from './styles';

const Skills = ({ skills }) => (
  <Sidebar>
    {skills.map(skillGroup => (
      <>
        {skillGroup.name === 'Frontend' ? (
          <FrontendSkillGroup key={skillGroup.name} details={skillGroup} />
        ) : (
          <SkillGroup key={skillGroup.name}>
            <SkillGroupTitle>{skillGroup.name}</SkillGroupTitle>
            <SkillsList>
              <SkillsLogo src={skillGroup.logo} alt="" />
              <SkillsGrid>
                {skillGroup.skills.map(skill => (
                  <SkillWrapper
                    key={skill.skill}
                    isLastElementAlone={skillGroup.skills.length % 2 !== 0}
                  >
                    <SkillName>{skill.skill}</SkillName>
                    <SkillLevel level={skill.level} />
                  </SkillWrapper>
                ))}
              </SkillsGrid>
            </SkillsList>
          </SkillGroup>
        )}
      </>
    ))}
  </Sidebar>
);

Skills.propTypes = {
  skills: arrayOf(skillGroupsPropType).isRequired
};

export default Skills;
