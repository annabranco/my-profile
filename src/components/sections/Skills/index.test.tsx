import { shallow, ShallowWrapper } from 'enzyme';
import 'jest-styled-components';
import Skills from '.';
import { setupMockProvider } from '../../../tests/mocks';

// TODO [10-May -21]: Implement defined testing suite (Anna Branco)

describe('< Skills >', () => {
  let MockProvider;
  let wrapper: ShallowWrapper;

  beforeAll(() => {
    MockProvider = setupMockProvider({ language: 'en' }).MockProvider;
    wrapper = shallow(
      <MockProvider>
        <Skills cuePointsActivated={new Set()} />
      </MockProvider>
    );
  });

  it('should mount', () => {
    expect(wrapper).toBeDefined();
  });

  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  xit('should pass information about Frontend skills on a specific component', () => {});

  xit('should render skills received from server separated by groups', () => {});

  xit('should render on each skill a SkillLevel subcomponent passing a skill level as props', () => {});

  xit('should display animation when cuepoints are activated', () => {});

  xit('should translate all texts when the language is changed', () => {});
});
