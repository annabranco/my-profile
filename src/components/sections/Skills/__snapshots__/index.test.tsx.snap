// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`< Skills > matches the snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Skills
    cuePointsActivated={Set {}}
  />
</Provider>
`;
